sentry_dsn_data.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/getsentry/sentry-go"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var DSNAnnotation = "k8s.sentry.io/dsn"
+
+// map from Sentry DSN to Client
+type DsnClientMapping struct {
+	mutex         sync.RWMutex
+	clientMap     map[string]*sentry.Client
+	customDsnFlag bool
+}
+
+func NewDsnData() *DsnClientMapping {
+	return &DsnClientMapping{
+		mutex:         sync.RWMutex{},
+		clientMap:     make(map[string]*sentry.Client),
+		customDsnFlag: isTruthy(os.Getenv("SENTRY_K8S_CUSTOM_DSNS")),
+	}
+}
+
+// return client if added successfully
+// (also returns client if already exists)
+func (d *DsnClientMapping) AddClientToMap(options sentry.ClientOptions) (*sentry.Client, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// check if we already encountered this dsn
+	existingClient, ok := d.clientMap[options.Dsn]
+	if ok {
+		return existingClient, nil
+	}
```

Comment:
Let's remove this check, IMO it's safer to assume that calling AddClientToMap will reinitialize the Client if the DSN already exists.